### PR TITLE
Add folder and record

### PR DESCRIPTION
**Pull request by cecilia-donnelly**
- **What kind of change does this PR introduce?**
This PR adds upload functionality which is present in the `3-upload-item` branch to `main`.  It is another step in the process started by PRs #10 and #11.  After this, there will be a final PR to bring share features to `main`.

- **What is the current behavior?**
I believe that currently uploads don't work on the main branch.

- **Other information**:
This is a WIP because the tests pass but I have not been able to make my credentials work to actually test uploads!  Hoping to get that sorted out in the next few hours.

### CODE REVIEW

**Comment by cecilia-donnelly** [comment]
Okay, @jasonaowen , this is no longer WIP!  I think it is ready to go.  To test it, I used a script that @andrewatwood left us, which is also available in our Google Doc.  I would like to include it in this repository or somewhere else accessible to anyone who wants to use the SDK, but I'm not sure the right way to do that.  What do you think?  It has a lot in common with the snippets in the README, naturally, but I found it easier to be able to run an existing script.

In the meantime, you might also like to use it.  I have included some inline comments about some things that tripped me up:

```
// I had to add this so that I could run the script against my local copy,
// which has an invalid cert.  Naturally you would not want to do this in
// real life.
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

// I used [npm link](https://docs.npmjs.com/cli/v7/commands/npm-link) to 
// make our in-progress package available locally, which was super easy
const permSdk = require('@permanentorg/node-sdk');
const Permanent = permSdk.Permanent;

// Pull these out of developer tools in any request in the web client.  Note
// that this is `archiveNbr`, not `archiveId`!  I was very confused by why
// this failed until I figured that out.  You don't have to use your local version
// (the SDK defaults to prod) but I found it easier to debug locally.
const permanent = new Permanent({
    'sessionToken': '',
    'mfaToken': '',
    'archiveNbr': '',
    'apiKey': '',
    'baseUrl': 'https://local.permanent.org/api',
});

// this is a very cute puppy, thank you Andrew.  Note that the
// SDK only accepts files that are already online somewhere.
const imageUrl = 'https://post.greatist.com/wp-content/uploads/sites/3/2020/02/322868_1100-1100x628.jpg';
const imageName = 'image.jpg';

run();

async function run() {
  const attempts = 1;

  await permanent.init();
  console.log('session valid!');

  const start = new Date();
  const folderForSet = await permanent.folder.create(`simultaneous test @ ${new Date().toISOString()}`);

  console.log('folder created');

  const logTime = () => {
    const end = new Date();
    const hours = (end.getTime() - start.getTime()) / 1000 / 60 / 60;
    console.log(`Time elapsed: ${hours} hours`);
  }

  try {
    for (let i = 0; i < attempts; i++) {
      await Promise.all([
        permanent.record.uploadFromUrl({displayName: 'image', uploadUri: imageUrl, uploadFileName: imageName}, folderForSet),
        permanent.record.uploadFromUrl({displayName: 'image', uploadUri: imageUrl, uploadFileName: imageName}, folderForSet),
      ]);
    }
    logTime();
    console.log('all success!');
  } catch (err) {
    console.dir(error);
    logTime();
    console.error('failure :(');
  }
}
```

**Review by jasonaowen — CHANGES_REQUESTED** [review]
Overall, the code looks good! I was able to successfully upload to dev using the test script. Thank you for taking the time to split this out from the larger branch, @cecilia-donnelly!

I think this pull request has a single, comprehensive change; after the initial commit, the subsequent commits seem to be patching up problems and adding missing pieces discovered via testing. That might be enough motivation to request a squash by itself, but I think a more compelling reason is that we should be crediting @andrewatwood! Would you mind squashing and adding a [Co-authored-by trailer](https://github.blog/2018-01-29-commit-together-with-co-authors/) to the commit message, please?

> I would like to include [this test script] in this repository or somewhere else accessible to anyone who wants to use the SDK, but I'm not sure the right way to do that. What do you think?

That sounds like a good idea to me! Maybe as `examples/upload.js` or so?

I do have some code review feedback on the test script itself.

Can we inline `Permanent`, so `const permanent = new permSdk.Permanent({ ...`?

That puppy picture is adorable! The original source seems to be from [Laula Co on Unsplash](https://unsplash.com/photos/sICtHYZzt_s). Unfortunately, I don't see very clear licensing terms on that page. I found several promising images on a [Creative Commons search for "puppy"](https://search.creativecommons.org/search?q=puppy&license=cc0,pdm); I particularly liked [Cute white puppy, by Wall Boat](https://www.flickr.com/photos/151415985@N06/23503731148) and [Wild Puppy, by Miwok](https://www.flickr.com/photos/11234074@N05/14013137587). Maybe we should use one of them if we're going to commit this script?

Why are we logging in hours rather than seconds? Why are there two `uploadFromUrl` calls? (I suspect the answer to both is that this was originally used for longer form testing; we should modify the code for the new use we're putting it to.)

Note the `catch (err)` / `console.dir(error)` mismatch, which we've talked about before.

**Comment by cecilia-donnelly** [comment]
@jasonaowen Okay, changes made!  Thanks for the thorough review.

File: examples/upload.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// To run this script, do `node upload.js`
```
Comment (jasonaowen) [inline examples/upload.js:1]: This file is currently marked executable. For it to be so, I think we need a hashbang, or else we get a bunch of errors like

```
./upload.js: line 1: //: Is a directory
./upload.js: line 2: //: Is a directory
./upload.js: line 3: //: Is a directory
./upload.js: line 4: //: Is a directory
./upload.js: line 5: //: Is a directory
./upload.js: line 7: syntax error near unexpected token `('
./upload.js: line 7: `const permSdk = require('@permanentorg/node-sdk');'
```

```suggestion
#!/usr/bin/env nodejs

// To run this script, do `node upload.js`
```

Happily, nodejs seems to know to ignore the hashbang. There's an interesting rabbit hole to fall down: [a proposal to make a leading hashbang part of the language](https://github.com/tc39/proposal-hashbang).

Anyway, the alternative is to not mark this file as executable, and leave the existing comment.

**Review by jasonaowen — APPROVED** [review]
Awesome! Thank you, @cecilia-donnelly.

Some small, optional comments on the example script; if you want to make those changes, go ahead and merge after doing so.

**Pull request merged**